control-plane/subcommand/common/common.go

```diff
@@ -88,29 +88,32 @@ func ValidateUnprivilegedPort(flagName, flagValue string) error {
 
 // ConsulLogin issues an ACL().Login to Consul and writes out the token to tokenSinkFile.
 // The logic of this is taken from the `consul login` command.
-func ConsulLogin(client *api.Client, cfg *api.Config, log hclog.Logger, bearerTokenFile, authMethodName, tokenSinkFile, namespace string, serviceAccountName string, meta map[string]string) error {
+func ConsulLogin(client *api.Client, cfg *api.Config, log hclog.Logger, bearerTokenFile, authMethodName, tokenSinkFile, namespace string, serviceAccountName string, meta map[string]string) (*api.ACLToken, error) {
 	// Read the bearerTokenFile.
 	data, err := ioutil.ReadFile(bearerTokenFile)
 	if err != nil {
-		return fmt.Errorf("unable to read bearerTokenFile: %v, err: %v", bearerTokenFile, err)
+		return nil, fmt.Errorf("unable to read bearerTokenFile: %v, err: %v", bearerTokenFile, err)
 	}
 	bearerToken := strings.TrimSpace(string(data))
 	if bearerToken == "" {
-		return fmt.Errorf("no bearer token found in %s", bearerTokenFile)
+		return nil, fmt.Errorf("no bearer token found in %s", bearerTokenFile)
 	}
+	var tok *api.ACLToken
 	err = backoff.Retry(func() error {
 		// Do the login.
 		req := &api.ACLLoginParams{
 			AuthMethod:  authMethodName,
 			BearerToken: bearerToken,
 			Meta:        meta,
 		}
-		tok, _, err := client.ACL().Login(req, &api.WriteOptions{Namespace: namespace})
+		tok, _, err = client.ACL().Login(req, &api.WriteOptions{Namespace: namespace})
 		if err != nil {
 			return fmt.Errorf("error logging in: %s", err)
 		}
 		// Write out the resultant token file.
-		if err := WriteFileWithPerms(tokenSinkFile, tok.SecretID, 0444); err != nil {
```

Comment:
I wonder if we should also make the writing to the file optional (maybe based on whether `tokenSinkFile` is provided or not). I'm thinking that when we're using it for consul client, we don't need the token to be written both to a file and to config json.

Reply:
I think you are also saying "Are we writing to a Sink file even though when this optional flag is not set?"???

Reply:
An additional question for context....Will consul agent merge the token in the sink file with the tokens specified in this type of config that is written on the client?
```
{
  "acl": {
    "enabled": true,
    "default_policy": "deny",
    "down_policy": "extend-cache",
    "tokens": {
      "agent": "{{ . }}"
    }
  }
```

Namely, we always need to write this part if `init-type="client"` because it not only sets the token but also enables ACls, sets deny as default, etc. ?

Reply:
>Are we writing to a Sink file even though when this optional flag is not set?

yep exactly. The consul client agent will only look at the token in the config and so writing it to a file as well is not necessary.

>Namely, we always need to write this part if init-type="client" because it not only sets the token but also enables ACls, sets deny as default, etc. ?

Yes.